File: addons/terrabrush/Scripts/Objects.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Godot;

namespace TerraBrush;

[Tool]
public partial class Objects : Node3D, IObjectsNode {
    private Texture2D _defaultNoise;
    private CancellationTokenSource _objectsCreationCancellationTokenSource;

    [Export] public int ObjectsIndex { get;set; }
    [Export] public ObjectDefinitionResource Definition { get;set; }
    [Export] public ZonesResource TerrainZones { get;set; }
    [Export] public int ZonesSize { get;set; }
    [Export] public float WaterFactor { get;set; }
    [Export] public bool LoadInThread { get;set; }
    [Export] public int DefaultObjectFrequency { get;set;}

    public override void _Ready() {
        base._Ready();

        if (Definition == null) {
            return;
        }

        _defaultNoise = ResourceLoader.Load<Texture2D>("res://addons/terrabrush/Resources/DefaultNoise.tres");

        UpdateObjects();
    }

    public override void _ExitTree() {
        base._ExitTree();

        _objectsCreationCancellationTokenSource?.Cancel();
    }

    private async void UpdateObjects() {
        if (LoadInThread) {
            _objectsCreationCancellationTokenSource?.Cancel();
            _objectsCreationCancellationTokenSource = new CancellationTokenSource();

            await Task.Factory.StartNew(async () => {
                await UpdateObjectsAsync(_objectsCreationCancellationTokenSource.Token);
            }, _objectsCreationCancellationTokenSource.Token);
        } else {
            await UpdateObjectsAsync(CancellationToken.None);
        }
    }

    private async Task UpdateObjectsAsync(CancellationToken cancellationToken) {
        if (cancellationToken.IsCancellationRequested) {
            return;
        }

        for (var zoneIndex = 0; zoneIndex < TerrainZones.Zones?.Length; zoneIndex++) {
            if (cancellationToken.IsCancellationRequested) {
                return;
            }

            var zone = TerrainZones.Zones[zoneIndex];

            var heightmapImage = zone.HeightMapTexture.GetImage();
            var waterImage = zone.WaterTexture?.GetImage();

            if (cancellationToken.IsCancellationRequested) {
                return;
            }

            var objectsContainerNode = new Node3D();
            objectsContainerNode.Name = $"{zoneIndex}";
            objectsContainerNode.Position = new Vector3(zone.ZonePosition.X * ZonesSize, 0, zone.ZonePosition.Y * ZonesSize);

            CallDeferred("add_child", objectsContainerNode);

            var imageTexture = zone.ObjectsTexture[ObjectsIndex];

            var noiseTexture = Definition?.NoiseTexture != null ? await Utils.WaitForTextureReady(Definition.NoiseTexture) : _defaultNoise;
            Image noiseImage = null;
            if (noiseTexture != null) {
                noiseImage = new Image();
                noiseImage.CopyFrom(noiseTexture.GetImage());
            }

            // Load all the objects from the image
            var objectsImage = imageTexture.GetImage();

            for (var x = 0; x < objectsImage.GetWidth(); x++) {
                if (cancellationToken.IsCancellationRequested) {
                    return;
                }

                for (var y = 0; y < objectsImage.GetHeight(); y++) {
                    if (cancellationToken.IsCancellationRequested) {
                        return;
                    }

                    var objectPixel = objectsImage.GetPixel(x, y);
                    CalculateObjectPresenceForPixel(
                        heightmapImage,
                        waterImage,
                        noiseImage,
                        x,
                        y,
                        objectPixel,
                        result => {
                            if (cancellationToken.IsCancellationRequested) {
                                return;
                            }

                            CallDeferred(
                                nameof(AddObjectNode),
                                objectsContainerNode,
                                $"{x}_{y}",
                                result.ResultPosition,
                                result.ResultRotation,
                                result.ResultPackedSceneIndex
                            );
                        }
                    );
                }
            }
        }
    }

    private void AddObjectNode(Node3D parentNode, string nodeName, Vector3 nodePosition, Vector3 nodeRotation, int packedSceneIndex) {
        var newNode =  Definition.ObjectScenes[packedSceneIndex].Instantiate<Node3D>();
        newNode.Name = nodeName;
        newNode.Position = nodePosition;
        newNode.RotationDegrees = nodeRotation;

        parentNode.AddChild(newNode);
    }

    public void UpdateObjectsHeight(List<ZoneResource> zones) {
        foreach (var zone in zones) {
            var zoneIndex = Array.IndexOf(TerrainZones.Zones, zone);
            var heightmapImage = zone.HeightMapTexture.GetImage();
            var waterImage = zone.WaterTexture?.GetImage();

            var nodeName = $"{zoneIndex}";
            var objectsNode = GetNodeOrNull(nodeName);

            if (objectsNode != null) {
                var noiseTexture = Definition.NoiseTexture ?? _defaultNoise;
                Image noiseImage = null;
                if (noiseTexture != null) {
                    noiseImage = noiseTexture.GetImage();
                }

                foreach (Node3D objectNode in objectsNode.GetChildren()) {
                    var objectNodeName = objectNode.Name;
                    var positions = objectNodeName.ToString().Split("_");
                    var xPosition = int.Parse(positions[0]);
                    var yPosition = int.Parse(positions[1]);

                    var resultPosition = GetPositionWithNoise(noiseImage, xPosition, yPosition);
                    var resultImagePosition = GetImagePosition(resultPosition.X, resultPosition.Z);
                    if (IsImagePositionInRange(resultImagePosition.X, resultImagePosition.Y)) {
                        objectNode.Position = new Vector3(objectNode.Position.X, GetObjectHeight(heightmapImage, waterImage, resultImagePosition.X, resultImagePosition.Y), objectNode.Position.Z);
                    }
                }
            }
        }
    }

    public void UpdateMeshesFromTool() {
        // Nothing to do here, stuff has already been applied
    }

    // TODO : Refactor this part so it shares the same code as the other strategy
    public void CalculateObjectPresenceForPixel(Image heightmapImage, Image waterImage, Image noiseImage, int x, int y, Color pixelValue, Action<(Vector3 ResultPosition, Vector3 ResultRotation, int ResultPackedSceneIndex)> objectPresentCallback, Action objectNotPresentCallback = null) {
        if (pixelValue.A > 0.0f) {
            var objectFrequency = Definition.ObjectFrequency < 1 ? DefaultObjectFrequency : Definition.ObjectFrequency;
            if (x % objectFrequency != 0 || y % objectFrequency != 0) {
                return;
            }

            var resultPosition = GetPositionWithNoise(noiseImage, x, y);
            var resultImagePosition = GetImagePosition(resultPosition.X, resultPosition.Z);
            if (IsImagePositionInRange(resultImagePosition.X, resultImagePosition.Y)) {
                var heightmapPixel = heightmapImage.GetPixel(resultImagePosition.X, resultImagePosition.Y);
                // Check for hole
                if (heightmapPixel.G == 0.0) {
                    var randomItemIndex = Utils.GetNextIntWithSeed((x * 1000) + y, 0, Definition.ObjectScenes.Count() - 1);
                    resultPosition -= new Vector3(ZonesSize / 2, -GetObjectHeight(heightmapImage, waterImage, resultImagePosition.X, resultImagePosition.Y), ZonesSize / 2);

                    objectPresentCallback(
                        (
                            resultPosition,
                            Definition.RandomYRotation ? new Vector3(0, Utils.GetNextFloatWithSeed((x * 1000) + y, 0f, 360f), 0) : Vector3.Zero,
                            randomItemIndex
                        )
                    );
                }
            }
        } else {
            objectNotPresentCallback?.Invoke();
        }
    }

    private Vector3 GetPositionWithNoise(Image noiseImage, int x, int y) {
        var resultPosition = new Vector3(x, 0, y);
        if (noiseImage != null) {
            var noisePixel = noiseImage.GetPixel(x, y).R;
            var randomValueX = Utils.GetNextFloatWithSeed((int) (noisePixel * 100), -Definition.RandomRange, Definition.RandomRange);
            var randomValueZ = Utils.GetNextFloatWithSeed(1 + (int) (noisePixel * 100), -Definition.RandomRange, Definition.RandomRange);
            resultPosition += new Vector3(randomValueX, 0, randomValueZ);
        }

        return resultPosition;
    }

    private Vector2I GetImagePosition(float x, float y) {
        return new Vector2I((int) Math.Round(x), (int) Math.Round(y));
    }

    private bool IsImagePositionInRange(int x, int y) {
        return x >= 0 && x < ZonesSize && y >= 0 && y < ZonesSize;
    }

    private float GetObjectHeight(Image heightmapImage, Image waterImage, int imageX, int imageY) {
        var heightmapPixel = heightmapImage.GetPixel(imageX, imageY);
        var waterHeight = waterImage?.GetPixel(imageX, imageY).R ?? 0;
        return heightmapPixel.R - (waterHeight * WaterFactor);
    }

    public void AddRemoveObjectFromTool(bool add, int x, int y, ZoneResource zone, Image heightmapImage, Image waterImage, Image noiseImage) {
        var zoneIndex = Array.IndexOf(TerrainZones.Zones, zone);
        var containerNode = GetNodeOrNull($"{zoneIndex}");
        if (containerNode == null) {
            containerNode = new Node3D() {
                Name = $"{zoneIndex}",
                Position = new Vector3(zone.ZonePosition.X * ZonesSize, 0, zone.ZonePosition.Y * ZonesSize)
            };
            AddChild(containerNode);
        }

        var nodeName = $"{x}_{y}";
        var existingNode = containerNode.GetNodeOrNull(nodeName);
        if (add && existingNode == null) {
            CalculateObjectPresenceForPixel(
                heightmapImage,
                waterImage,
                noiseImage,
                x,
                y,
                Colors.White,
                result => {
                    CallDeferred(
                        nameof(AddObjectNode),
                        containerNode,
                        nodeName,
                        result.ResultPosition,
                        result.ResultRotation,
                        result.ResultPackedSceneIndex
                    );
                }
            );
        } else if (!add && existingNode != null) {
            existingNode.QueueFree();
        }
    }
}

File: src/nodes/objects.cpp
#include "objects.h"
#include "../misc/zone_utils.h"
#include "../misc/utils.h"

#include <godot_cpp/classes/resource_loader.hpp>
#include <godot_cpp/classes/image.hpp>
#include <godot_cpp/classes/texture2d.hpp>
#include <godot_cpp/classes/packed_scene.hpp>

using namespace godot;

void Objects::_bind_methods() {
    ClassDB::bind_method(D_METHOD("updateObjectsAsync"), &Objects::updateObjectsAsync);
    ClassDB::bind_method(D_METHOD("addObjectNode", "parentNode", "nodeName", "nodePosition", "nodeRotation", "nodeSizeFactor", "packedSceneIndex"), &Objects::addObjectNode);
}

void Objects::_notification(int what) {
    switch (what) {
        case NOTIFICATION_EXIT_TREE: {
            if (_objectsThread.is_valid()) {
                _objectsCreationCancellationTokenSource.cancel();
                _objectsThread->wait_to_finish();
            }

            _objectsThread.unref();
        } break;
    }
}

Objects::Objects() {
    _defaultObjectFrequency = 10;
}

Objects::~Objects() {}

void Objects::_ready() {
    if (_definition.is_null()) {
        return;
    }

    _defaultNoise = ResourceLoader::get_singleton()->load("res://addons/terrabrush/Resources/DefaultNoise.tres");

    updateObjects();
}

void Objects::updateObjects() {
    if (_loadInThread) {
        if (_objectsThread.is_valid()) {
            _objectsCreationCancellationTokenSource.cancel();
            _objectsThread->wait_to_finish();
        }

        _objectsCreationCancellationTokenSource = CancellationSource();

        _objectsThread.instantiate();
        _objectsThread->start(Callable(this, "updateObjectsAsync"));
    }
    else {
        updateObjectsAsync();
    }
}

void Objects::updateObjectsAsync() {
    CancellationToken cancellationToken = _objectsCreationCancellationTokenSource.token;

    if (cancellationToken.isCancellationRequested) {
        return;
    }

    for (int zoneIndex = 0; zoneIndex < _terrainZones->get_zones().size(); zoneIndex++) {
        if (cancellationToken.isCancellationRequested) {
            return;
        }

        Ref<ZoneResource> zone = _terrainZones->get_zones()[zoneIndex];

        Ref<Image> heightmapImage = zone->get_heightMapTexture()->get_image();
        Ref<Image> waterImage;

        if (!zone->get_waterTexture().is_null()) {
            waterImage = zone->get_waterTexture()->get_image();
        }

        if (cancellationToken.isCancellationRequested) {
            return;
        }

        Node3D *objectsContainerNode = memnew(Node3D);
        objectsContainerNode->set_name(String::num_int64(zoneIndex));
        objectsContainerNode->set_position(Vector3(zone->get_zonePosition().x * _zonesSize, 0, zone->get_zonePosition().y * _zonesSize));

        call_deferred("add_child", objectsContainerNode);

        Ref<Texture2D> imageTexture = zone->get_objectsTexture()[_objectsIndex];

        Ref<Texture2D> noiseTexture;
        if (_definition->get_noiseTexture().is_null()) {
            noiseTexture = _defaultNoise;
        } else {
            noiseTexture = _definition->get_noiseTexture();
        }

        Ref<Image> noiseImage;
        if (!noiseTexture.is_null()) {
            noiseImage = Ref<Image>(memnew(Image));
            noiseImage->copy_from(noiseTexture->get_image());
        }

        // Load all the objects from the image
        Ref<Image> objectsImage = imageTexture->get_image();

        for (int x = 0; x < objectsImage->get_width(); x++) {
            if (cancellationToken.isCancellationRequested) {
                return;
            }

            for (int y = 0; y < objectsImage->get_height(); y++) {
                if (cancellationToken.isCancellationRequested) {
                    return;
                }

                Color objectPixel = objectsImage->get_pixel(x, y);
                calculateObjectPresenceForPixel(
                    objectsContainerNode,
                    heightmapImage,
                    waterImage,
                    noiseImage,
                    x,
                    y,
                    objectPixel
                );
            }
        }
    }
}

void Objects::addObjectNode(const Node3D *parentNode, const String nodeName, const Vector3 nodePosition, const Vector3 nodeRotation, const float nodeSizeFactor, const int packedSceneIndex) {
    Ref<PackedScene> objectPackedScene = _definition->get_objectScenes()[packedSceneIndex];
    Node3D *newNode = (Node3D*)(objectPackedScene->instantiate());

    newNode->set_name(nodeName);
    newNode->set_position(nodePosition);
    newNode->set_rotation_degrees(nodeRotation);
    newNode->set_scale(newNode->get_scale() * nodeSizeFactor);

    const_cast<Node3D*>(parentNode)->add_child(newNode);
}

void Objects::calculateObjectPresenceForPixel(Node3D *parentNode, Ref<Image> heightmapImage, Ref<Image> waterImage, Ref<Image> noiseImage, int x, int y, Color pixelValue) {
    if (pixelValue.a > 0.0f) {
        int objectFrequency = _definition->get_objectFrequency() < 1 ? _defaultObjectFrequency : _definition->get_objectFrequency();
        if (x % objectFrequency != 0 || y % objectFrequency != 0) {
            return;
        }

        Vector3 resultPosition = getPositionWithNoise(noiseImage, x, y);
        if (isImagePositionInRange(resultPosition.x, resultPosition.z)) {
            int resolutionZoneSize = ZoneUtils::getImageSizeForResolution(_zonesSize, _resolution);
            Vector2 heightImagePosition = getHeightPositionForResolution(Vector2(resultPosition.x, resultPosition.z), resolutionZoneSize);
            Color heightmapPixel = Utils::getPixelLinear(heightmapImage, heightImagePosition.x, heightImagePosition.y);
            // Check for hole
            if (heightmapPixel.g == 0.0) {
                int randomItemIndex = Utils::getNextIntWithSeed((x * 1000) + y, 0, _definition->get_objectScenes().size() - 1);
                resultPosition -= Vector3(_zonesSize / 2, -getObjectHeight(heightmapImage, waterImage, heightImagePosition.x, heightImagePosition.y), _zonesSize / 2);

                Vector3 resultRotation = _definition->get_randomYRotation() ? Vector3(0, Utils::getNextFloatWithSeed((x * 1000) + y, 0, 360), 0) : Vector3(0, 0, 0);
                float resultSizeFactor = _definition->get_randomSize() ? Utils::getNextFloatWithSeed((x * 1000) + y, _definition->get_randomSizeFactorMin(), _definition->get_randomSizeFactorMax()) : 1.0f;

                call_deferred("addObjectNode", parentNode, String(String::num_int64(x) + "_" + String::num_int64(y)), resultPosition, resultRotation, resultSizeFactor, randomItemIndex);
            }
        }
    }
}

Vector3 Objects::getPositionWithNoise(Ref<Image> noiseImage, int x, int y) {
    int remapXPixel = (int) Math::round(Math::remap((float) x, 0, _zonesSize - 1, 0, noiseImage->get_width() - 1));
    int remapYPixel = (int) Math::round(Math::remap((float) y, 0, _zonesSize - 1, 0, noiseImage->get_height() - 1));

    Vector3 resultPosition = Vector3(x, 0, y);
    if (!noiseImage.is_null()) {
        float noisePixel = noiseImage->get_pixel(remapXPixel, remapYPixel).r;
        float randomValueX = Utils::getNextFloatWithSeed((int)(noisePixel * 100), -_definition->get_randomRange(), _definition->get_randomRange());
        float randomValueZ = Utils::getNextFloatWithSeed(1 + (int)(noisePixel * 100), -_definition->get_randomRange(), _definition->get_randomRange());
        resultPosition += Vector3(randomValueX, 0, randomValueZ);
    }

    return resultPosition;
}

bool Objects::isImagePositionInRange(float x, float y) {
    return x >= 0.0 && x < _zonesSize && y >= 0.0 && y < _zonesSize;
}

float Objects::getObjectHeight(Ref<Image> heightmapImage, Ref<Image> waterImage, float imageX, float imageY) {
    Color heightmapPixel = Utils::getPixelLinear(heightmapImage, imageX, imageY);
    float waterHeight = waterImage.is_null() ? 0 : Utils::getPixelLinear(waterImage, imageX, imageY).r;
    return heightmapPixel.r - (waterHeight * _waterFactor);
}

Vector2 Objects::getHeightPositionForResolution(Vector2 position, int resolutionZoneSize) {
    if (_resolution != 1) {
        return Vector2(
            Math::remap(position.x, 0, _zonesSize - 1, 0, resolutionZoneSize - 1),
            Math::remap(position.y, 0, _zonesSize - 1, 0, resolutionZoneSize - 1)
        );
    }

    return position;
}

void Objects::updateObjectsHeight(TypedArray<Ref<ZoneResource>> zones) {
    int resolutionZoneSize = ZoneUtils::getImageSizeForResolution(_zonesSize, _resolution);

    for (Ref<ZoneResource> zone : zones) {
        int zoneIndex = _terrainZones->get_zones().find(zone);
        Ref<Image> heightmapImage = zone->get_heightMapTexture()->get_image();
        Ref<Image> waterImage;
        if (!zone->get_waterTexture().is_null()) {
            waterImage = zone->get_waterTexture()->get_image();
        }

        String nodeName = String::num_int64(zoneIndex);
        Node *objectsNode = get_node_or_null(nodeName);

        if (objectsNode != nullptr) {
            Ref<Texture2D> noiseTexture = _definition->get_noiseTexture().is_null() ? _defaultNoise : _definition->get_noiseTexture();
            Ref<Image> noiseImage;
            if (!noiseTexture.is_null()) {
                noiseImage = noiseTexture->get_image();
            }

            for (int i = 0; i < objectsNode->get_child_count(); i++) {
                Node3D *objectNode = (Node3D*)(objectsNode->get_child(i));
                String objectNodeName = objectNode->get_name();
                PackedStringArray positions = objectNodeName.split("_");
                int xPosition = positions[0].to_int();
                int yPosition = positions[1].to_int();

                Vector3 resultPosition = getPositionWithNoise(noiseImage, xPosition, yPosition);
                if (isImagePositionInRange(resultPosition.x, resultPosition.z)) {
                    Vector2 heightImagePosition = getHeightPositionForResolution(Vector2(resultPosition.x, resultPosition.z), resolutionZoneSize);
                    objectNode->set_position(Vector3(objectNode->get_position().x, getObjectHeight(heightmapImage, waterImage, heightImagePosition.x, heightImagePosition.y), objectNode->get_position().z));
                }
            }
        }
    }
}

void Objects::updateMeshesFromTool() {
    // Nothing to do here, stuff has already been applied
}

void Objects::addRemoveObjectFromTool(bool add, int x, int y, Ref<ZoneResource> zone, Ref<Image> heightmapImage, Ref<Image> waterImage, Ref<Image> noiseImage) {
    int zoneIndex = _terrainZones->get_zones().find(zone);
    Node3D *containerNode = (Node3D*)(get_node_or_null(String::num_int64(zoneIndex)));
    if (containerNode == nullptr) {
        containerNode = memnew(Node3D);
        containerNode->set_name(String::num_int64(zoneIndex));
        containerNode->set_position(Vector3(zone->get_zonePosition().x * _zonesSize, 0, zone->get_zonePosition().y));
        add_child(containerNode);
    }

    String nodeName = String(String::num_int64(x) + "_" + String::num_int64(y));
    Node3D *existingNode = (Node3D*)(containerNode->get_node_or_null(nodeName));
    if (add && existingNode == nullptr) {
        calculateObjectPresenceForPixel(
            containerNode,
            heightmapImage,
            waterImage,
            noiseImage,
            x,
            y,
            Color::named("WHITE")
        );
    }
    else if (!add && existingNode != nullptr) {
        existingNode->queue_free();
    }
}
